contracts/accessless/OlympusChainlinkBasedSwapper.sol

```diff
@@ -0,0 +1,215 @@
+pragma solidity ^0.8.7;
+
+import "@chainlink/contracts/src/v0.8/interfaces/FeedRegistryInterface.sol";
+import "@chainlink/contracts/src/v0.8/Denominations.sol";
+import "@uniswap/v3-periphery/contracts/interfaces/ISwapRouter.sol";
+import "@uniswap/v2-periphery/contracts/interfaces/IUniswapV2Router02.sol";
+import "@openzeppelin/contracts/proxy/utils/Initializable.sol";
+import "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+import "@openzeppelin/contracts/token/ERC20/extensions/IERC20Metadata.sol";
+
+struct V3Params {
+    uint24[] fees;
+    address[] path;
+    address denomination;
+    address recipient;
+    uint256 deadline;
+    uint256 amount;
+    uint256 slippage;
+}
+
+struct V2Params {
+    address router;
+    address[] path;
+    address denomination;
+    address recipient;
+    uint256 deadline;
+    uint256 amount;
+    uint256 slippage;
+}
+
+error OlympusChainlinkBasedSwapper_WrongV3PathLength();
+error OlympusChainlinkBasedSwapper_ApproveFailed();
+error OlympusChainlinkBasedSwapper_TransferFromFailed();
+error OlympusChainlinkBasedSwapper_TransferFailed();
+
+contract OlympusChainlinkBasedSwapper {
+    address public constant weth = 0xC02aaA39b223FE8D0A0e5C4F27eAD9083C756Cc2;
+    address public constant wbtc = 0x2260FAC5E5542a773Aa44fBCfeDf7C193bc2C599;
+
+    FeedRegistryInterface public registry;
+    ISwapRouter public v3SwapRouter;
+
+    constructor(address _registry, address _v3SwapRouter) {
+        registry = FeedRegistryInterface(_registry);
+        v3SwapRouter = ISwapRouter(_v3SwapRouter);
+    }
+
+    function v3ExactInput(V3Params calldata params) external returns (uint256 amountOut) {
+        amountOut = _exact(params, true);
+    }
+
+    function v3ExactOutput(V3Params calldata params) external returns (uint256 amountIn) {
+        amountIn = _exact(params, false);
+    }
+
+    function v2Swap(V2Params calldata params) external {
+        // reads
+        address inputToken = params.path[0];
+        address outputToken = params.path[params.path.length - 1];
+        IUniswapV2Router02 router = IUniswapV2Router02(params.router);
+        address denomination = (params.denomination == address(0)) ? Denominations.USD : params.denomination;
+
+        // interactions + checks
+        // didn't abstract to keep readable
+
+        bool check = IERC20(inputToken).transferFrom(msg.sender, address(this), params.amount);
+
+        if (!check) revert OlympusChainlinkBasedSwapper_TransferFromFailed();
+
+        check = IERC20(inputToken).approve(params.router, params.amount);
```

Comment:
does approve work on objects like this rather than addresses? very good to know

Reply:
what? IERC20(inputToken) == IERC20, params.router == address, amount == uint256